Add medkits, remove apples

ClobberXD opened the pull request:
- **Apples have been removed in favour of _medkits_**.
- On use, medkits provide 1 HP every 0.5s to the player till a maximum of `max_regen` is attained.
- Players will be slowed down while they heal, and their screen will have slight bluish tint, representing the player's lack of attention to the surroundings while healing.
- Healing players will also glow to visually distinguish them from those who aren't healing.
- If healing is interrupted, the regen will stop, the one medkit will be added back to player's inventory and their health will be reverted back to it's original value.
  - Healing can be interrupted in the following ways:
    - Healing player engages in PvP, or gets attacked by another player.
    - Any other damage to healing player.

**Tested; everything works as it should.**
 Video demonstration: https://youtu.be/wfYTRcP170A, https://youtu.be/o86Y7aKHP1Q

rubenwardy commented on mods/medkit/init.lua:
```diff
@@ -0,0 +1,115 @@
+-- ID of "healing effect" HUD element
+local players = {}
+
+local on_use
```
I'd prefer the functions to be defined like  `local function on_use(`  and above the register_item

ClobberXD replied:
That's what I was going for initially, but `local function update_hp` isn't called when recursively invoked from within `update_hp` itself. Should I remove `local` from these functions

rubenwardy commented on mods/medkit/init.lua:
```diff
@@ -0,0 +1,115 @@
+		speed = 0.5
+	})
+
+	minetest.after(0.5, update_hp(user))
```
```lua
minetest.after(0.5, update_hp, user)
```

ClobberXD replied:
oops

rubenwardy commented on mods/medkit/init.lua:
```diff
@@ -0,0 +1,115 @@
+-- Called on left-click and is recursively called
+-- every 0.5s for 3 seconds after left-click
+update_hp = function(player)
+	local name = player:get_player_name()
```
should just stop here if nothing in players

rubenwardy commented on mods/medkit/init.lua:
```diff
@@ -0,0 +1,115 @@
+	local name = player:get_player_name()
+	player:set_hp(player:get_hp() + 1)
+
+	if players[name] and os.time() < players[name].time and
```
please do

```lua
local data = players[name]
```

and use that here

rubenwardy commented on mods/medkit/init.lua:
```diff
@@ -0,0 +1,115 @@
+	end
+
+	-- Remove one health pack from player's inv
+	player:get_inventory():remove_item("main", Itemstack("medkit:medkit 1"))
```
what if they have dropped the item?

ClobberXD replied:
Yes, I'm in the process of adding an on_drop callback already :)

rubenwardy commented on mods/medkit/init.lua:
```diff
@@ -0,0 +1,115 @@
+-- Reset effects
+reset_effects = function(player)
+	user:hud_remove(players[name].hud)
+	user:set_physics_override()
```
does this actually work? It may need to be `{ key = 1}` instead

ClobberXD replied:
Yet to test, but that should definitely work to the best of my knowledge

ClobberXD replied:
True, this results in weird behaviour. My bad

rubenwardy commented:
Also, I don't like the texture. Something like this would be better:

![](https://rimworldwiki.com/images/2/27/Medicine.png)

(note that the above is copyrighted)

ClobberXD commented:
Updated major portions. See first comment for more details

AKryukov92 commented:
Just random idea about medkits.
![image](https://user-images.githubusercontent.com/24266616/45338832-0c008680-b5b1-11e8-8737-38496de145e9.png)

You can also think about "apple jam" instead of apples themselves.

ClobberXD commented:
While apples have indeed played an unforgettable part in CTF's development as a server, they result in gameplay that's far from realistic. Apples (and all food, in general) are out-of-place in the middle of a battlefield, and proper medical supplies (like medkits, bandages, etc.) adhere to the theme of CTF (i.e. Warfare) nicely.

AKryukov92 commented:
I don't like the idea of cloning some other ctf. It doesn't really matter, how exactly we will name healing stuff. May be you should ask community? Something like:
```
We are going to change healing in the game. New item will heal over time. How to name it?
* Apple
* Medkit
* Bandage
```

Btw, bandages for healing others is good idea, but they are not subject of this topic.

ClobberXD commented:
What should be the behaviour of medkits with respect to sprint? Should speed be half of sprint when player is sprinting?

afshaan4 commented:
I think they should slow you down to an absolute speed, not a relative speed.

ClobberXD commented:
But that wouldn't respect sprint, would it?

I'm thinking of adding a `medkit.is_healing(name)` which sprint uses to set a physics modifier accordingly, since sprint already registers a globalstep.

AKryukov92 commented:
Why do you want to allow players to move while healing?

ClobberXD commented:
In a fast-paced game, I think it's overkill to completely disallow players to move.

AKryukov92 commented:
Will player with healing and spring be faster than normal running speed?

ClobberXD commented:
No. It is (supposed to be) **half** of normal and sprint speeds

AKryukov92 commented:
If you fight 1vs1 and you want to heal, then you need to run away faster than opponent (hoping that he has less sprint available). You need to run first and heal later.
If you fight many vs many, you have a chance to run back and heal while your teammates are fighting. It doesn't really matter if you will be able to run or not. Your mates will cover you. It is better to run first and heal later anyway.

In addition, I would like to keep feature "guns can interrupt healing from the distance". It makes guns more useful that they are now.

ClobberXD commented:
> If you fight 1vs1 and you want to heal, then you need to run away faster than opponent (hoping that he has less sprint available). You need to run first and heal later.

It's not intended to be used in the middle of a fight. The primary objective of medkits is to make healing relatively more realistic. Healing while fighting is very unrealistic, and medkits aims to disallow players from doing that.
> I would like to keep feature "guns can interrupt healing from the distance". It makes guns more useful that they are now.

I quite agree on this. It would be useful if there is an `on_damage` callback available...

AKryukov92 commented:
> Healing while fighting is very unrealistic, and medkits aims to disallow players from doing that.

Healing while running is not very realistic either.

afshaan4 commented:
Which is exactly why you can't run while healing

ClobberXD commented:
**Major update**
- Sprint is now completely disabled while healing.
- I've dropped the `on_drop` callback, as I realized that there was no need for one. One medkit is used up immediately after player left-clicks, and dropping all the (remaining) medkits should not interrupt the healing in any way. This was an unforeseen overthought on my part.

I've also added some testing code which reduces the player's health and gives them 10 medkits. This will make it easy for anyone who would like to test this PR themselves. I'll remove this code when the PR is ready to merge.

ClobberXD commented:
Updated texture. Added video demonstration to first post.

ClobberXD commented on mods/random_messages/init.lua:
```diff
@@ -48,7 +48,8 @@ end
 	random_messages.messages = {
 		"To talk to only your team, start your messages with /t. For example, /t Hello team!",
-		"Eat apples to restore health quickly.",
+		"Medkits will gradually restore health, at the rate of half a heart every second.",
```
Needs updating, thanks for pointing that out

ClobberXD replied:
Now that the update interval is configurable, any idea as to what the random message should be?

AKryukov92 replied:
I propose "Medkits will restore half of a heart every $regen_interval second" but English is not my native language, so there might be mistakes.

ClobberXD replied:
It's not possible to access `regen_interval` from within `random_messages/init.lua`. So I was thinking of a more generic message by completely omitting the "every x seconds" part.

AKryukov92 replied:
What about "Use medkits in safe place. Healing will slow you down."

ClobberXD replied:
Sounds good

ClobberXD replied:
```lua
"Use medkits to gradually restore your health.",
"Using medkits will slow you down, and you won't be able to sprint while healing.",
```

ClobberXD replied:
The working of medkits is a little more complex than apples, and compressing all of that into one random message would make it too long. Hence, I added another message, just for the slowing effect.

ClobberXD commented on mods/medkit/init.lua:
```diff
@@ -0,0 +1,118 @@
+	if interrupted then
+		minetest.chat_send_player(name, minetest.colorize("#FF4444",
+				"Your healing was interrupted!"))
+		player:set_hp(players[name].hp - max_regen)
```
Let's assume player abc uses a medkit. A table is added to `players`, indexed by `abc`. Here's what it looks like if player uses medkit when at 16 HP:

```lua
players[abc] = {
    hp = 22, -- Assuming max_regen = 6, hp = 16 + 6 = 22
    hud = <hud_element_def>
}
```

If healing is interrupted, player's original hp is restored by `player:set_hp(players[abc].hp - max_regen)`. Since `players[abc].hp` is 22, subtracting `max_regen` (6) produces the original HP of abc, 22 - 6 = 16.

ClobberXD commented on mods/medkit/init.lua:
```diff
@@ -0,0 +1,118 @@
+
+-- DEBUG
+
+local function debug(player)
```
I was planning on removing these before merging, after rubenwardy approves the PR. Should I remove it right now?

ClobberXD commented on mods/medkit/init.lua:
```diff
@@ -0,0 +1,118 @@
+-- Recursively called after left-click every 0.5s for a
+-- total regen of max_regen, limited by player's max hp (20)
+function update_hp(player)
+	local name = player:get_player_name()
```
That escaped my attention... nice find!

ClobberXD replied:
Added an `on_leaveplayer` callback

ClobberXD commented on mods/medkit/init.lua:
```diff
@@ -0,0 +1,118 @@
+-- stop regen and revert back to original state
+minetest.register_on_player_hpchange(function(player, hp, reason)
+	-- Check if player takes damage
+	if hp < 0 then
```
`hp` holds the actual HP difference - a negative value (`hp < 0`) indicates damage, and a positive value indicates healing.

Also, haven't yet tested interruption with a teammate. Will do.

ClobberXD replied:
Interruption by teammate does not seem to be possible, as the damage is not even registered. An `on_punchplayer` callback disables damage if both players are on the same team:

```lua
if to.team == from.team and to.team ~= "" and to.team ~= nil and to.name ~= from.name then
	minetest.chat_send_player(hitter:get_player_name(), player:get_player_name() .. " is on your team!")
	if not ctf.setting("friendly_fire") then
		return true
	end
end
```

ClobberXD commented on mods/medkit/init.lua:
```diff
@@ -0,0 +1,118 @@
+	end
+	local name = user:get_player_name()
+
+	if players[name] or user:get_hp() == 20 then
```
I dropped the magic number in favour of Object Property `hp_max`, but there's an issue here. For some reason, `hp_max` has a value of 10 (i.e. 5 hearts) only.

@rubenwardy Should I set `hp_max` to 20 within an `on_joinplayer` callback?

ClobberXD commented on mods/medkit/init.lua:
```diff
@@ -0,0 +1,118 @@
+	user:set_properties({glow = 3})
+	user:set_physics_override({speed = 0.5})
+
+	minetest.after(0.5, update_hp, user)
```
Done

AKryukov92 commented on mods/medkits/init.lua:
```diff
@@ -0,0 +1,135 @@
+
+-- Boolean function for use by other mods to check if a player is healing
+function is_healing(name)
+	return players[name] and true or false
```
Is there some lua-magic with `nil`s and booleans?
Statement `condition and true` is equivalent to `condition`.
Statement `condition or false` is equivalent to `condition`.
Why don't you just write `return players[name] != nil`?

ClobberXD replied:
Well, this is a ternary operator of sorts, and it works just as well.

It evaluates to `true` if `players[name]` exists, and `false` otherwise.

AKryukov92 commented on mods/medkits/init.lua:
```diff
@@ -0,0 +1,135 @@
+	players[name] = nil
+end
+
+-- Recursively called after left-click every n seconds (determined by
```
It is not recursive anymore.

AKryukov92 commented on mods/medkits/init.lua:
```diff
@@ -0,0 +1,135 @@
+	on_use = start_healing
+})
+
+--
```
I think you should remove them now because when rubenwardy approves PR, you should not do any additional changes.
General rule: if you add commit (or replace commit), PR should be reviewed again.
If you want to keep those functions, then it is better to move them to special separate mod.

ClobberXD commented:
Some extensive testing is necessary as this is a major change. I don't trust myself to test out the PvP-related code in this mod correctly, as PvP between two clients from the exact same machine is very buggy.

AKryukov92 commented:
Finally got to my computer and did some extensive testing on this PR.

Subject: *Apples have been removed in favour of medkits.*
Testing case: Load Two lakes and two bridges and look around
Expect no apples on small trees.

Subject: *On use, medkits provide 1 HP every 0.5s to the player till a maximum of max_regen is attained.*
Testing case: Decrease hp to 6 hearts. start healing. wait for healing to finish. Check amount of HP.
Expect HP to be equal 11.
Correct.

Testing case: Decrease hp to 6 hearts. Start healing. Wait for healing to finish. Start healing again. Wait for healing to finish. Check amount of HP.
Expect HP to be equal 17.
Wrong. Sometimes I can't start healing second time.
My assumption: When player uses medkit first, server stores his max hp value and blocks healing when player has more tan this amount of hp.

Subject: *Players will be slowed down while they heal, and their screen will have slight bluish tint, representing the player's lack of attention to the surroundings while healing.*
Testing case: Decrease hp to 6 hearts. Start healing. Move around longer than healing duration.
Expect slowed speed during healing and normal speed after.
Correct.

Subject: *Players will be slowed down while they heal, and their screen will have slight bluish tint, representing the player's lack of attention to the surroundings while healing.*
Testing case: Decrease hp to 6 hearts. Start healing. Watch out for screen tint.
Expect screen to become blue and return to normal color when healing is finished.
Correct.

Testing case: *Decrease hp to 6 hearts. Take hit so the screen becomes red. Use medkit. Look for screen tint.*
Expect screen to become blue and return to normal color when healing is finished.
Correct.

Subject: *Healing players will also glow to visually distinguish them from those who aren't healing.*
Testing case: Decrease hp to 6 hearts. Start healing. Look from other player point of view.
Expect something around healing player or healing color color change.
Wrong. Healing player is no different from anyone else.

Subject: *If healing is interrupted, the regen will stop, the one medkit will be added back to player's inventory and their health will be reverted back to it's original value.
Healing can be interrupted in the following ways:
Healing player engages in PvP, or gets attacked by another player.
Any other damage to healing player.*
Testing case: Decrease hp to 6 hearts. Start healing. Attack during healing.
Expect healing to be interrupted. Blue blur should fade. Speed should return to normal.
Wrong. Server crashed.
```
2018-11-03 22:30:06: ERROR[Main]: stack traceback:
2018-11-03 22:30:06: ERROR[Main]:       [C]: in function 'set_hp'
2018-11-03 22:30:06: ERROR[Main]:       ...server\bin\..\games\capturethefla
g\mods\medkits\init.lua:48: in function 'stop_healing'
2018-11-03 22:30:06: ERROR[Main]:       ...server\bin\..\games\capturethefla
g\mods\medkits\init.lua:86: in function 'func'
2018-11-03 22:30:06: ERROR[Main]:       ...t-0.4.17.1-win64-server\bin\..\bu
iltin\game\register.lua:511: in function <...t-0.4.17.1-win64-server\bin\..\
builtin\game\register.lua:507>
```
My assumption is that "reverting hitpoints" triggers "register_on_player_hpchange" and you have got indirect recursion.

I hope it will be helpful.

DS-Minetest commented on mods/medkits/init.lua:
```diff
@@ -0,0 +1,118 @@
+	regen_timer = regen_timer + dtime
+	if regen_timer < regen_interval then
+		return
+	end
```
It might be a bit better using `minetest.after`.
See discussion at https://github.com/minetest-mods/item_drop/commit/e84433d41ccd4abec1795acbcebe2895bc5be380#diff-d7d74284e2e39f6c06d199c8e2bedbd0.

ClobberXD replied:
If you reach the end of that discussion, it gets established that minetest.after does not perform worse than globalstep, but neither does it perform better - after all, minetest.after itself is created from globalstep.

ClobberXD commented on mods/medkits/init.lua:
```diff
@@ -0,0 +1,118 @@
+
+	player:set_physics_override({speed = 1})
+	player:hud_remove(players[name].hud)
+	players[name] = nil
```
What's the reason for this suggestion?

DS-Minetest replied:
The following happens when the player `player` with the medikit-stack `stack` and with the name `name` heals and gets damaged:

step | `players[name] ~= nil` | function call | description
--- | ---- | --- | -----
1 | `false` -> `true` | `start_healing(stack, player)` | a player uses a medkit
2 | `true` | `register_on_player_hpchange…` | the player gets damage
3 | `true` | `stop_healing(player, true)` | `players[name] ~= nil` is `true`, so interrupt healing
4 | `true` | `player:set_hp(players[name].hp_target - regen_max)` | the healing was interrupted, so set the hp back
5 **= 2** | `true` | `register_on_player_hpchange…` | the player gets damage

`not players[name]` must be `true` when `set_hp` is called, else you get a never ending loop, which is bad.

@AKryukov92 reported this, it caused a crash for him.

ClobberXD commented:
@AKryukov92 Thank you for the extensive testing and feedback. I have removed the glow effect, by the way. But I'll look into the server crash when player engages in PvP while healing. Again, thank you very much for the detailed testing. :+1:

AKryukov92 commented:
While I'm agree with new mechanic of healing, I don't understand why you are replacing apple item.
Apple could heal just like medkit.
Healing with apples is rather unique idea, which makes MT-CTF very different from all other CTF-games.
You can find medkits (or healing potions) almost everywhere and only apples are present in MT-CTF.

ClobberXD commented:
> Testing case: Decrease hp to 6 hearts. Start healing. Wait for healing to finish. Start healing again. Wait for healing to finish. Check amount of HP.
Expect HP to be equal 17.
Wrong. Sometimes I can't start healing second time.
My assumption: When player uses medkit first, server stores his max hp value and blocks healing when player has more tan this amount of hp.

@AKryukov92 This is because the hp_max property of all the players default to a value of `10`, and using medkits when having >10 HP results in the healing being aborted. I'm thinking of manually setting a value of 20 for all players on_join...

Thomas--S commented:
> hp_max property of all the players default to a value of 10

As far as I know, HP are twice the number of hearts, so the default max value of hp_max is 20 anyway.

ClobberXD commented:
> As far as I know, HP are twice the number of hearts, so the default max value of hp_max is 20 anyway.

That's exactly what I assumed. But then I used `print()` to verify this, and voila - `hp_max = 10`. It is supposed to be 20 by default. So I assume that some mod is interfering with this. I'll get to the bottom of this soon.

AKryukov92 commented:
Subject: *Apples have been removed in favour of medkits.*
Testing case: Load Two lakes and two hills and look around
Expect no apples on small trees.
Correct.

Subject: *On use, medkits provide 1 HP every 0.5s to the player till a maximum of max_regen is attained.*
Testing case: Decrease hp to 8½ hearts (17 hp). Start healing. wait for healing to finish. Check amount of HP.
Expect HP to be 10 hearts, 20 hp.
Wrong. Healing didn't started.

Testing case: Decrease hp to 4 hearts (8 hp). start healing. wait for healing to finish. Check amount of hp.
Expect 7 hearts (14 hp).
Wrong. When healing has finished, player had 5 hearts (10 hp).

Testing case: Decrease hp to 1 heart (2 hp). start healing. wait for healing to finish. Check amount of hp.
Expect 4 hearts (8 hp)
Correct.

Subject: *Players will be slowed down while they heal, and their screen will have slight bluish tint, representing the player's lack of attention to the surroundings while healing.*
Testing case: Decrease hp to 4 hearts (8 hp). Start healing. Move around longer than healing duration.
Expect slowed speed during healing and normal speed after.
Correct. Even if healing ended at 5 hearts, speed was restored correctly.

Testing case: Decrease hp to 1 heart (2 hp). Start healing. Move around longer than healing duration.
Expect slowed speed during healing and normal speed after.
Correct.

Subject: *Players will be slowed down while they heal, and their screen will have slight bluish tint, representing the player's lack of attention to the surroundings while healing.*
Testing case: Decrease hp to 4 hearts (8 hp). Start healing. Watch out for screen tint.
Expect screen to become blue and return to normal color when healing is finished.
Correct. Even if healing ended at 5 hearts, screen color was restored to normal.

Testing case: Decrease hp to 1 heart (2 hp). Take hit so the screen becomes red. Use medkit. Look for screen tint.
Expect screen to become blue and return to normal color when healing is finished.
Correct.2544

Subject: *If healing is interrupted, the regen will stop, the one medkit will be added back to player's inventory and their health will be reverted back to it's original value.
Healing can be interrupted in the following ways:
Healing player engages in PvP, or gets attacked by another player.
Any other damage to healing player.*
Testing case: Decrease hp to 1 heart. Start healing. Attack during healing with melee weapon.
Expect healing to be interrupted. Blue blur should fade. Speed should return to normal.
Correct. Healing was interrupted. 1 medkit was returned to healed player.

Testing case: Decrease hp to 1 heart. Start healing. Attack during healing with crossbow
Expect healing to be interrupted. Blue blur should fade. Speed should return to normal.
Correct. Healing was interrupted. 1 medkit was returned to healed player.

Testing case: Decrease hp to 1 heart. Start healing. Attack during healing with smg.
Expect healing to be interrupted. Blue blur should fade. Speed should return to normal.
Correct. Healing was interrupted. 1 medkit was returned to healed player.

Testing case: Decrease hp to 1 heart. Start healing. Take damage by falling.
Expect healing to be interrupted. Blue blur should fade. Speed should return to normal.
Correct. Healing was interrupted. 1 medkit was returned to healed player.

Testing case: Decrease hp to 1 heart. Start healing. Take damage by lava.
Expect healing to be interrupted. Blue blur should fade.
Correct. Healing was interrupted. 1 medkit was returned to healed player.

Testing case: Decrease hp to 1 heart. Start healing. Attack during healing with sword and kill.
Expect player death and all his items dropped on ground.
Correct. Healing was interrupted. Amount of dropped medkits was equal to the amount before healing.

ClobberXD commented:
Since we only want to get rid of apple's healing property, I've re-added apples, as they could continue existing in maps as cosmetic elements.

I've also made them `walkable`, not sure about this, though.

AKryukov92 commented:
CTF version 50e64e343c8ef774dbaa1aa84853a4e8d3b96e96

I was too lazy to check all test cases, because second test case is incorrect.

Subject: Apples exists on trees, but can't be eaten.
Testing case: Load Two lakes and two hills and look around
Expect apples on small trees.
Correct.

Subject: On use, medkits provide 1 HP every 0.5s to the player till a maximum of max_regen is attained.
Testing case: Decrease hp to 8½ hearts (17 hp). Start healing. wait for healing to finish. Check amount of HP.
Expect HP to be 10 hearts, 20 hp.
Wrong. Healing didn't started.

I still propose to give new healing mechanic to old apple items. I mean `on_use = start_healing` for apples.

ClobberXD commented:
Healing stops at 10 HP due to the `hp < hp_max` check, which returns false when hp == 10 because of the issue of `hp_max` strangely being initialized to a value of 10 as I reported in one of my previous comments. But thanks for testing again.

AKryukov92 commented:
Why do you use `tonumber(player:get_properties().hp_max)` instead of `20`?
It looks hacky, but I'm afraid this is something from MT core and we probably can't fix it in CTF so easily.

ClobberXD commented:
But `hp_max` should default to 20 according to the engine docs. If it's 10, then something else is interfering, but I can't seem to find any instance of `hp_max` being set to 10 in the whole repo.

ClobberXD commented:
- All references to `hp_max` property has been dropped in favour of a hard-value of 20.
- The medkits texture has been improved by quite a bit (thanks @ChandlerBing1994)

AKryukov92 commented:
Subject: On use, medkits provide 1 HP every 0.5s to the player till a maximum of max_regen is attained.
Testing case: Decrease hp to 8½ hearts (17 hp). Start healing. wait for healing to finish. Check amount of HP.
Expect HP to be 10 hearts, 20 hp.
Correct.

Testing case: Decrease hp to 4 hearts (8 hp). start healing. wait for healing to finish. Check amount of hp.
Expect 7 hearts (14 hp).
Correct.

Testing case: Decrease hp to 1 heart (2 hp). start healing. wait for healing to finish. Check amount of hp.
Expect 4 hearts (8 hp)
Correct.

Subject: Players will be slowed down while they heal, and their screen will have slight bluish tint, representing the player's lack of attention to the surroundings while healing.
Testing case: Decrease hp to 4 hearts (8 hp). Start healing. Move around longer than healing duration.
Expect slowed speed during healing and normal speed after.
Correct.

Testing case: Decrease hp to 1 heart (2 hp). Start healing. Move around longer than healing duration.
Expect slowed speed during healing and normal speed after.
Correct.

Subject: Players will be slowed down while they heal, and their screen will have slight bluish tint, representing the player's lack of attention to the surroundings while healing.
Testing case: Decrease hp to 4 hearts (8 hp). Start healing. Watch out for screen tint.
Expect screen to become blue and return to normal color when healing is finished.
Correct.

Testing case: Decrease hp to 1 heart (2 hp). Take hit so the screen becomes red. Use medkit. Look for screen tint.
Expect screen to become blue and return to normal color when healing is finished.
Correct.

Subject: If healing is interrupted, the regen will stop, the one medkit will be added back to player's inventory and their health will be reverted back to it's original value.
Healing can be interrupted in the following ways:
Healing player engages in PvP, or gets attacked by another player.
Any other damage to healing player.
Testing case: Decrease hp to 1 heart. Start healing. Attack during healing with melee weapon.
Expect healing to be interrupted. Blue blur should fade. Speed should return to normal.
Correct. Healing was interrupted. 1 medkit was returned to healed player.

Testing case: Decrease hp to 1 heart. Start healing. Attack during healing with crossbow
Expect healing to be interrupted. Blue blur should fade. Speed should return to normal.
Correct. Healing was interrupted. 1 medkit was returned to healed player.

Testing case: Decrease hp to 1 heart. Start healing. Attack during healing with smg.
Expect healing to be interrupted. Blue blur should fade. Speed should return to normal.
Correct. Healing was interrupted. 1 medkit was returned to healed player.

Testing case: Decrease hp to 1 heart. Start healing. Take damage by falling.
Expect healing to be interrupted. Blue blur should fade. Speed should return to normal.
Correct. Healing was interrupted. 1 medkit was returned to healed player.

Testing case: Decrease hp to 1 heart. Start healing. Take damage by lava.
Expect healing to be interrupted. Blue blur should fade.
Correct. Healing was interrupted. 1 medkit was returned to healed player.

Testing case: Decrease hp to 1 heart. Start healing. Attack during healing with sword and kill.
Expect player death and all his items dropped on ground.
Correct. Healing was interrupted. Amount of dropped medkits was equal to the amount before healing.

AKryukov92 commented:
I stongly suggest to make a poll on the forum when this pull request will be merged and applied to live server. Something like:
* "Have you been killed by someone who uses medkits?"
* "Does medkit saved your life in battle?"

rubenwardy commented:
The point of medkits is to limit the healing power and make healing hacks useless. This could be done by changing how apples work instead of medkits. The difference between them is only really where they are found

So I don't mind whether this code is attached to apples or a new medkit item. As said - the only difference is that apples are spawned on trees. It may be worth gathering public opinion on this

AKryukov92 commented:
> It may be worth gathering public opinion on this

It is better to try and revert, than endlessly discuss it with public.

MinetestSam commented:
Maybe reduce the hitpoints healed by apples?

ClobberXD commented:
> Maybe reduce the hitpoints healed by apples?

To clarify: Are you suggesting apples remain alongside medkits, @MinetestSam?

Apples would either be removed completely, being replaced by medkits, or they would take up the healing mechanism that this PR provides.

MinetestSam commented:
Well then I have no problem to apples being removed

Thomas--S commented:
> Apples would either be removed completely, being replaced by medkits, or they would take up the healing mechanism that this PR provides.

I think that the second option might make it easier for new players to understand how this game works because apples are related to healing in most minetest subgames.

AKryukov92 commented:
I support the option "apples should take up the new healing mechanism" because healing apples are distinctive feature of MT-CTF. It makes this game original, compared to numerous other, non-MT ctf games.

MinetestSam commented:
Since @rubenwardy wants to promote teamwork then what about bandages from the Persistent Kingdoms server? The bandages when used only heals the teammates and not the person using it so this will promote more of teammates helping each other out

ClobberXD commented:
There are plans to allow players to heal other players using medkits.

MinetestSam commented:
Rather than implementing all the feature through medkits, I think it will be better to separate it like in games like PUBG, Fortnite, H1Z1 there are bandages and then there also are medkits both items are valuable and have there importance. Medkits heals you for full HP and bandages heal for 75% of the HP

Thomas--S commented:
How about using apples (with the proposed new mechanism) for self-healing and medkits for healing team mates?

ClobberXD commented:
I just realized that it's not possible for a player to manually interrupt healing, without engaging in PvP or falling on purpose. I consider it essential to be able to abort healing if the player senses danger, and has to get out of harm's way before being able to heal.

> I support the option "apples should take up the new healing mechanism" because healing apples are distinctive feature of MT-CTF.

> I think [apples with the new healing mechanism] might make it easier for new players to understand how this game works because apples are related to healing in most minetest subgames.

One drawback is that players would be _more_ confused, if they eat an apple expecting to immediately regain 1 heart, but are subjected to various other effects instead. From that point of view, I think a new item altogether would help players adapt to the new healing system better, without the extra chaos.

AKryukov92 commented:
I think it is better to merge medkits as they are and only then add new functionality.
Medkits are done. Why do you want to delay their release for indefinite amount of time?

MinetestSam commented:
I don't think adding multiple items that have a bit of difference will have any negative impact on CTF in fact it would give players more options to choose from and it will also make the healing mechanism a bit team dependent too

ClobberXD commented:
> I think it is better to merge medkits as they are and only then add new functionality. Medkits are done. Why do you want to delay their release for indefinite amount of time?

The game will be unplayable if deployed in its current state, as players won't be able to manually stop healing without taking damage. I prefer to complete the feature, and achieve the least possible amount of playability, rather than adding a broken feature.

> I don't think adding multiple items that have a bit of difference will have any negative impact on CTF in fact it would give players more options to choose from and it will also make the healing mechanism a bit team dependent too

While I'm not against multiple healing items, I just don't feel the necessity for more than one item. That being said, we could add them later, if required, once the players get used to the new healing system.

ClobberXD commented:
Updated the PR to interrupt healing if player moves more than 1m in any direction. Ready for testing/review.

rubenwardy commented:
7d396d2f6099788840bf177352777691958a6dd4
